lib/Dialect/FIRRTL/FIRRTLOps.cpp

```diff
@@ -1593,69 +1570,86 @@ FIRRTLType DShlPrimOp::getResultType(FIRRTLType lhs, FIRRTLType rhs,
   return IntType::get(lhs.getContext(), lhsi.isSigned(), width);
 }
 
-FIRRTLType DShlwPrimOp::getResultType(FIRRTLType lhs, FIRRTLType rhs,
-                                      Location loc) {
+INFER_BINARY_RETURN_TYPE(DShlwPrimOp) {
   if (!lhs.isa<IntType>() || !rhs.isa<UIntType>()) {
-    mlir::emitError(loc,
-                    "first operand should be integer, second unsigned int");
+    if (loc)
+      mlir::emitError(*loc,
+                      "first operand should be integer, second unsigned int");
     return {};
   }
   return lhs;
 }
 
-FIRRTLType DShrPrimOp::getResultType(FIRRTLType lhs, FIRRTLType rhs,
-                                     Location loc) {
+INFER_BINARY_RETURN_TYPE(DShrPrimOp) {
   if (!lhs.isa<IntType>() || !rhs.isa<UIntType>()) {
-    mlir::emitError(loc,
-                    "first operand should be integer, second unsigned int");
+    if (loc)
+      mlir::emitError(*loc,
+                      "first operand should be integer, second unsigned int");
     return {};
   }
   return lhs;
 }
 
+#undef INFER_BINARY_RETURN_TYPE
+
 //===----------------------------------------------------------------------===//
 // Unary Primitives
 //===----------------------------------------------------------------------===//
 
-FIRRTLType AsSIntPrimOp::getResultType(FIRRTLType input, Location loc) {
+template <class ConcreteOp>
+static inline FIRRTLType inferUnaryReturnType(FIRRTLType, Optional<Location>);
+
+#define INFER_UNARY_RETURN_TYPE(CLASS)                                         \
```

Comment:
It looks like an issue in this patch is that some places (unary / binary ops) you're adjusting the interface between the super generic "inferReturnTypes" interface and the more specific to firrtl interface given we only have one result, no regions, never needs the context, etc. 

Have you considered pushing the interface goop up a level into the tblgen level?  You could make each class that participates in this implement a semi-generic signature like:

FIRRTLType Yourclass::inferFIRRTLReturnTypes(ArrayRef<Type> operandTypes, NamedAttrList Attrs, OPtional<Location> loc) 

And then poop out something into the header like this:
LogicalResult inferReturnTypes(MLIRContext *context, Optional<Location> loc, ValueRange operands,       
     DictionaryAttr attrs, RegionRange regions,                             
     SmallVectorImpl&lt;Type&gt; &amp;results) { 
  return implementFIRRTLReturnInference(...all the operands..., &inferFIRRTLReturnTypes);
}

Then put implementFIRRTLReturnInference in some c++ file.  This would make the adaptation code common to subfield, mux etc.

Reply:
it would also allow the parser to call into `inferFIRRTLReturnTypes` which would save creating temporary DictionaryAttrs etc.

Reply:
@lattner That was my first reflex as well, but I couldn't figure out how to keep TableGen from emitting the `inferReturnTypes(<many-args>)` prototype. The C++ compiler then complains about the decl `inferReturnTypes(<many-args>);` colliding with the decl+def `inferReturnTypes(<many-args>) { ... }`. There's an option in `DeclareOpInterfaceMethods` that allows you to *always* emit a function decl in case you want to override a default, but we'd need the inverse to prevent it from emitting the decl since we do so on our own.

Am I missing something here?

Reply:
ok, I don't know how this specific interface works :).  We can also work with the upstream project (as you're doing already) to help improve things if needbe.  I'm sure we're not the only ones that would benefit from a simpler interface